kingpin/actors/packagecloud.py

```diff
@@ -0,0 +1,358 @@
+
+https://packagecloud.io/docs/api
+
+**Required Environment Variables**
+
+:PACKAGECLOUD_ACCOUNT:
+  packagecloud account name, i.e. https://packagecloud.io/PACKAGECLOUD_ACCOUNT
+
+:PACKAGECLOUD_TOKEN:
+  packagecloud API Token
+"""
+
+import datetime
+import logging
+import os
+import re
+import sys
+
+from tornado import gen
+from tornado_rest_client import api
+
+from kingpin.actors import base
+from kingpin.actors import exceptions
+from kingpin.constants import REQUIRED
+
+log = logging.getLogger(__name__)
+
+__author__ = 'Charles McLaughlin <charles@nextdoor.com>'
+
+ACCOUNT = os.getenv('PACKAGECLOUD_ACCOUNT', None)
+TOKEN = os.getenv('PACKAGECLOUD_TOKEN', None)
+
+
+class PackagecloudAPI(api.RestConsumer):
+
+    ENDPOINT = 'https://packagecloud.io/api/v1/'
+    CONFIG = {
+        'attrs': {
+            'packages': {
+                'path': ('repos/%account%/%repo%/packages.json'
+                         '?per_page={}'.format(sys.maxint)),
+                'http_methods': {'get': {}}
+            },
+            'delete': {
+                'path': 'repos/%account%/%repo%/%distro_version%/%filename%',
+                'http_methods': {'delete': {}}
+            },
+        },
+        'auth': {
+            'user': TOKEN,
+            'pass': ''
+        }
+    }
+
+
+class PackagecloudBase(base.HTTPBaseActor):
+
+    """Simple packagecloud Abstract Base Object"""
+
+    def __init__(self, *args, **kwargs):
+        """Check required environment variables."""
+        super(PackagecloudBase, self).__init__(*args, **kwargs)
+
+        if not ACCOUNT:
+            raise exceptions.InvalidCredentials(
+                'Missing the "PACKAGECLOUD_ACCOUNT" environment variable.')
+
+        if not TOKEN:
+            raise exceptions.InvalidCredentials(
+                'Missing the "PACKAGECLOUD_TOKEN" environment variable.')
+
+        rest_client = api.RestClient()
+        self._packagecloud_client = PackagecloudAPI(client=rest_client)
+
+    @gen.coroutine
+    def _get_all_packages(self, repo):
+        """Simple method for fetching a dictionary of all packages in a repo"""
+        all_packages = yield self._packagecloud_client.packages(
+            token=TOKEN, account=ACCOUNT, repo=repo).http_get()
+        raise gen.Return(all_packages)
+
+    @gen.coroutine
+    def _delete(self, packages_to_delete, repo, older_than=0,
+                number_to_keep=0):
+        """Generic packagecloud delete method, optionally supporting deleting
+        old packages by date."""
+
+        deleted_packages = []
+        all_packages = yield self._get_all_packages(repo=repo)
+
+        # Extract a list of unique package names
+        package_names = [package['name'] for package in all_packages]
+        package_names = set(package_names)
+
+        pattern = re.compile(packages_to_delete)
+
+        # Loop through each unique package
+        for name in package_names:
+            if not pattern.match(name):
```

Comment:
Rather than looping like this and using `continue` to skip packages we're not deleting, instead create the list of packages to delete ahead of time. Something like:

```
deletable_packages = [package['name'] in all_packages if pattern.match(package['name'])]

for package in deletable_packages:
  ...
```